### PR TITLE
Make Trainer evaluation handle dynamic seq_length

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

In token classification problems or language modeling predictions, logits and labels are usually of shape `[batch_size, seq_length, ...]` and if one uses dynamic padding, the `seq_length` might change from batch to batch. This made the evaluate and predict methods of the `Trainer` fail, because the Trainer wants to concatenate all predictions in one array.

This PR fixes this, by potentially padding up to the maximum seq length when needed, using the -100 pad index (since it's the marker for ignored labels in PyTorch). It add tests that were failing before the PR, to check the fix actually solves the issue.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1333,6 +1333,12 @@ def predict(self, test_dataset: Dataset) -> PredictionOutput:
                 Dataset to run the predictions on. If it is an :obj:`datasets.Dataset`, columns not accepted by the
                 ``model.forward()`` method are automatically removed. Has to implement the method :obj:`__len__`
 
+        .. note::
+
+            If your predictions or labels have different sequence length (for instance because you're doing dynamic
```
Comment (LysandreJik) [inline src/transformers/trainer.py:1338]:
```suggestion
            If your predictions or labels have different sequence lengths (for instance because you're doing dynamic
```

File: src/transformers/trainer_pt_utils.py [hunk 2]
```diff
@@ -42,17 +42,50 @@
 logger = logging.get_logger(__name__)
 
 
-def nested_concat(tensors, new_tensors, dim=0):
-    "Concat the `new_tensors` to `tensors` on `dim`. Works for tensors or nested list/tuples of tensors."
+def torch_pad_and_concatenate(tensor1, tensor2, pad_idx=-100):
```
Comment (LysandreJik) [inline src/transformers/trainer_pt_utils.py:45]: Can the argument be `padding_id`, as that's the wording used elsewhere? (this applies to all methods used here)

**Review by LysandreJik — APPROVED** [review]
Looks good to me!
